File: src/research1.js
require('dotenv').config();
const fs = require('fs');
const path = require('path');
const { xfetch, delay, saveCSV, loadJSON } = require('./xfetch-client');

const HANDLES_FILE = './data/research1-handles.json';
const ENRICHED_CSV = './data/my-follows-enriched.csv';
const OUT_DIR = './data/research1';
const DAYS_BACK = 15;

function parseEnrichedCSV() {
  const raw = fs.readFileSync(ENRICHED_CSV, 'utf-8');
  const lines = raw.trim().split('\n');
  const headers = lines[0].split(',');
  return lines.slice(1).map(line => {
    const vals = [];
    let current = '';
    let inQuotes = false;
    for (const ch of line) {
      if (ch === '"') { inQuotes = !inQuotes; }
      else if (ch === ',' && !inQuotes) { vals.push(current.trim()); current = ''; }
      else { current += ch; }
    }
    vals.push(current.trim());
    const obj = {};
    headers.forEach((h, i) => { obj[h.trim()] = vals[i] || ''; });
    return obj;
  });
}

function analyzeSentiment(text) {
  const lower = (text || '').toLowerCase();
  const bullish = ['bullish', 'breakout', 'rally', 'green', 'long', 'buy', 'strength',
    'upside', 'rip', 'moon', 'pump', 'ath', 'new high', 'upgrad', 'beat', 'outperform',
    'accumulat', 'add', 'conviction', 'bounce', 'oversold', 'dip buying'];
  const bearish = ['bearish', 'breakdown', 'sell', 'red', 'short', 'weak', 'downside',
    'dump', 'crash', 'correction', 'downgrad', 'miss', 'underperform', 'distribution',
    'caution', 'risk-off', 'defensive', 'overbought', 'top', 'recession', 'tariff'];

  let b = 0, be = 0;
  for (const w of bullish) if (lower.includes(w)) b++;
  for (const w of bearish) if (lower.includes(w)) be++;

  if (b > be) return 'bullish';
  if (be > b) return 'bearish';
  return 'neutral';
}

function buildTimeline(tweets) {
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - DAYS_BACK);

  const daily = {};
  for (const t of tweets) {
    const dt = t.created_at ? new Date(t.created_at) : null;
    if (!dt || isNaN(dt.getTime()) || dt < cutoff) continue;
    const key = dt.toISOString().slice(0, 10);
    if (!daily[key]) daily[key] = { bullish: 0, bearish: 0, neutral: 0, total: 0, tweets: [] };
    daily[key][t.sentiment]++;
    daily[key].total++;
    daily[key].tweets.push({
      handle: t.handle,
      text: (t.text || '').slice(0, 150),
      sentiment: t.sentiment,
      time: dt.toISOString().slice(11, 16),
    });
  }

  return Object.keys(daily).sort().map(date => {
    const d = daily[date];
    const score = d.total === 0 ? 50 : ((d.bullish - d.bearish) / d.total) * 50 + 50;
    return {
      date,
      bullish: d.bullish,
      bearish: d.bearish,
      neutral: d.neutral,
      total: d.total,
      sentiment_score: Math.round(score * 10) / 10,
      sentiment_label: score > 60 ? 'bullish' : (score < 40 ? 'bearish' : 'neutral'),
      sample_tweets: d.tweets.slice(0, 3).map(t => `@${t.handle}: ${t.text}`),
    };
  });
}

(async () => {
  console.log('[research1] xfetch-powered sentiment analysis\n');

  const handles = JSON.parse(fs.readFileSync(HANDLES_FILE, 'utf-8'));
  console.log(`  Targets: ${handles.map(h => '@' + h).join(', ')}`);
  console.log(`  Last ${DAYS_BACK} days\n`);

  // Resolve handles to IDs via xfetch (these are discovered accounts, not in enriched CSV)
  const idMap = {};
  for (const h of handles) {
    console.log(`  Resolving @${h}...`);
    const res = await xfetch(`/users/by-username/${h}`);
    idMap[h] = res.data.id;
  }

  // Phase 1: Collect tweets
  const allTweets = [];
  for (const handle of handles) {
    const uid = idMap[handle];
    console.log(`  [${handle}] Fetching tweets (id: ${uid})...`);

    try {
      const res = await xfetch(`/users/${uid}/tweets`);
      const tweets = res.data || [];
      const charged = res.meta?.credits?.charged || 0;
      const remaining = res.meta?.credits?.remaining || 0;

      for (const t of tweets) {
        allTweets.push({
          handle,
          tweet_id: t.id,
          created_at: t.created_at,
          text: t.text || '',
          likes: t.like_count || 0,
          retweets: t.retweet_count || 0,
          replies: t.reply_count || 0,
          is_retweet: t.is_retweet,
          is_reply: t.is_reply,
          sentiment: null,
        });
      }
      console.log(`    ${tweets.length} tweets (credits: -${charged}, ${remaining} left)`);
    } catch (e) {
      console.log(`    Error: ${e.message}`);
    }
    await delay(500);
  }

  console.log(`\n  Total: ${allTweets.length} tweets`);

  // Phase 2: Sentiment analysis
  console.log('  Classifying sentiment...');
  for (const t of allTweets) {
    t.sentiment = analyzeSentiment(t.text);
  }
  const sentiments = {};
  for (const t of allTweets) sentiments[t.sentiment] = (sentiments[t.sentiment] || 0) + 1;
  console.log(`    Bullish: ${sentiments.bullish || 0} | Bearish: ${sentiments.bearish || 0} | Neutral: ${sentiments.neutral || 0}`);

  // Phase 3: Timeline
  console.log('  Building timeline...');
  const timeline = buildTimeline(allTweets);

  // Phase 4: Save
  if (!fs.existsSync(OUT_DIR)) fs.mkdirSync(OUT_DIR, { recursive: true });

  // Tweets CSV
  saveCSV(
    path.join(OUT_DIR, 'tweets.csv'),
    allTweets.map(t => [t.handle, t.created_at, t.text.replace(/[\n\r]/g, ' '), t.sentiment, t.likes, t.retweets, t.replies, t.is_retweet ? 'yes' : 'no']),
    ['handle', 'created_at', 'text', 'sentiment', 'likes', 'retweets', 'replies', 'is_retweet'],
  );

  // Timeline CSV
  saveCSV(
    path.join(OUT_DIR, 'timeline.csv'),
    timeline.map(t => [t.date, t.bullish, t.bearish, t.neutral, t.total, t.sentiment_score, t.sentiment_label]),
    ['date', 'bullish', 'bearish', 'neutral', 'total', 'sentiment_score', 'sentiment_label'],
  );

  console.log('\n' + '='.repeat(60));
  console.log('  SENTIMENT TIMELINE');
  console.log('='.repeat(60));
  for (const t of timeline) {
    const bar = '█'.repeat(Math.round(t.sentiment_score / 5));
    const label = t.sentiment_label === 'bullish' ? '🟢' : t.sentiment_label === 'bearish' ? '🔴' : '⚪';
    console.log(`  ${t.date}  ${label} ${t.sentiment_label.toUpperCase().padEnd(8)}  score:${String(t.sentiment_score).padStart(5)}  ${bar}`);
    for (const s of t.sample_tweets.slice(0, 2)) {
      console.log(`            ${s.slice(0, 100)}`);
    }
  }

  console.log(`\n  Outputs: ${OUT_DIR}/tweets.csv, ${OUT_DIR}/timeline.csv`);
})();

File: src/research1.py
"""research1: Top trading accounts → 15-day tweets → sentiment timeline"""
import os, sys, json, csv
from datetime import datetime, timedelta, timezone
from collections import defaultdict

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from Scweet import Scweet

BASE_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..")
OUT_DIR = os.path.join(BASE_DIR, "data", "research1")
HANDLES_FILE = os.path.join(BASE_DIR, "data", "research1-handles.json")
COOKIES_FILE = os.path.join(BASE_DIR, "data", "cookies.json")
DAYS_BACK = 15

os.makedirs(OUT_DIR, exist_ok=True)

def load_handles(path):
    with open(path) as f:
        return json.load(f)

def parse_date(ts):
    if not ts:
        return None
    ts = str(ts).strip()
    formats = [
        "%a %b %d %H:%M:%S %z %Y",
        "%a %b %d %H:%M:%S %Y",
        "%Y-%m-%dT%H:%M:%S.%fZ",
        "%Y-%m-%dT%H:%M:%SZ",
        "%Y-%m-%d %H:%M:%S",
    ]
    for fmt in formats:
        try:
            return datetime.strptime(ts, fmt)
        except:
            continue
    return None

def analyze_sentiment(tweets):
    bullish = ['bullish', 'breakout', 'rally', 'green', 'long', 'buy', 'strength',
               'upside', 'rip', 'moon', 'pump', 'ATH', 'new high', 'upgrad',
               'beat', 'outperform', 'accumulat', 'add', 'conviction']
    bearish = ['bearish', 'breakdown', 'sell', 'red', 'short', 'weak', 'downside',
               'dump', 'crash', 'correction', 'downgrad', 'miss', 'underperform',
               'distribution', 'caution', 'risk-off', 'defensive', 'overbought']

    for t in tweets:
        text = (t.get('text', '') or '').lower()
        b_count = sum(1 for w in bullish if w in text)
        be_count = sum(1 for w in bearish if w in text)
        if b_count > be_count:
            t['sentiment'] = 'bullish'
        elif be_count > b_count:
            t['sentiment'] = 'bearish'
        else:
            t['sentiment'] = 'neutral'
    return tweets

def build_timeline(tweets):
    daily = defaultdict(lambda: {'bullish': 0, 'bearish': 0, 'neutral': 0, 'total': 0, 'tweets': []})
    cutoff = datetime.now(timezone.utc) - timedelta(days=DAYS_BACK)

    for t in tweets:
        dt = parse_date(t.get('created_at'))
        if not dt:
            continue
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        if dt < cutoff:
            continue

        date_key = dt.strftime("%Y-%m-%d")
        sent = t.get('sentiment', 'neutral')
        daily[date_key][sent] += 1
        daily[date_key]['total'] += 1
        daily[date_key]['tweets'].append({
            'handle': t.get('handle', ''),
            'text': (t.get('text', '') or '')[:200],
            'sentiment': sent,
            'time': dt.strftime("%H:%M"),
        })

    timeline = []
    for date_key in sorted(daily.keys()):
        d = daily[date_key]
        total = d['total']
        if total == 0:
            score = 50
        else:
            score = ((d['bullish'] - d['bearish']) / total) * 50 + 50
        timeline.append({
            'date': date_key,
            'bullish': d['bullish'],
            'bearish': d['bearish'],
            'neutral': d['neutral'],
            'total': total,
            'sentiment_score': round(score, 1),
            'sentiment_label': 'bullish' if score > 60 else ('bearish' if score < 40 else 'neutral'),
            'sample_tweets': [d['tweets'][i]['text'][:120] for i in range(min(3, len(d['tweets'])))],
        })

    return timeline

def main():
    handles = load_handles(HANDLES_FILE)
    since_date = (datetime.now(timezone.utc) - timedelta(days=DAYS_BACK)).strftime("%Y-%m-%d")

    print(f"[research1] {len(handles)} target accounts: {', '.join('@'+h for h in handles)}")
    print(f"  Since: {since_date}")

    s = Scweet(
        cookies_file=COOKIES_FILE,
        db_path=os.path.join(OUT_DIR, "scweet_state.db"),
    )

    # Phase 1: Collect tweets via search (get_profile_tweets is broken in v5.3)
    all_tweets = []
    for i, handle in enumerate(handles):
        print(f"\n  [{i+1}/{len(handles)}] @{handle}")
        try:
            results = s.search(f"from:{handle}", since=since_date, limit=500)
            for t in results:
                t['handle'] = handle
                t['created_at'] = t.get('timestamp', '')
                t['text'] = t.get('text', '') or ''
                t['likes'] = t.get('likes', 0)
            all_tweets.extend(results)
            print(f"    Collected {len(results)} tweets")
        except Exception as e:
            print(f"    Error: {e}")

    print(f"\n  Total tweets collected: {len(all_tweets)}")

    # Phase 2: Sentiment analysis
    print("  Running sentiment analysis...")
    all_tweets = analyze_sentiment(all_tweets)

    # Phase 3: Build timeline
    print("  Building sentiment timeline...")
    timeline = build_timeline(all_tweets)

    # Phase 4: Save outputs
    csv_path = os.path.join(OUT_DIR, "tweets.csv")
    with open(csv_path, 'w', newline='', encoding='utf-8') as f:
        writer = csv.DictWriter(f, fieldnames=['handle', 'created_at', 'text', 'sentiment', 'likes', 'retweets', 'replies'])
        writer.writeheader()
        for t in all_tweets:
            dt = parse_date(t.get('created_at'))
            writer.writerow({
                'handle': t.get('handle', ''),
                'created_at': dt.isoformat() if dt else '',
                'text': (t.get('text', '') or '').replace('\n', ' ') if t.get('text') else '',
                'sentiment': t.get('sentiment', ''),
                'likes': t.get('likes', 0) or t.get('like_count', 0),
                'retweets': t.get('retweets', 0) or t.get('retweet_count', 0),
                'replies': t.get('comments', 0) or t.get('reply_count', 0),
            })
    print(f"  Saved tweets → {csv_path}")

    tl_path = os.path.join(OUT_DIR, "timeline.csv")
    with open(tl_path, 'w', newline='', encoding='utf-8') as f:
        writer = csv.DictWriter(f, fieldnames=['date', 'bullish', 'bearish', 'neutral', 'total', 'sentiment_score', 'sentiment_label'])
        writer.writeheader()
        for t in timeline:
            writer.writerow({
                'date': t['date'],
                'bullish': t['bullish'],
                'bearish': t['bearish'],
                'neutral': t['neutral'],
                'total': t['total'],
                'sentiment_score': t['sentiment_score'],
                'sentiment_label': t['sentiment_label'],
            })
    print(f"  Saved timeline → {tl_path}")

    # Print summary
    print(f"\n  {'='*60}")
    print(f"  SENTIMENT TIMELINE (last {len(timeline)} days)")
    print(f"  {'='*60}")
    for t in timeline:
        bar = '█' * int(t['sentiment_score'] / 5)
        label = {'bullish': '🟢', 'bearish': '🔴', 'neutral': '⚪'}.get(t['sentiment_label'], '?')
        print(f"  {t['date']}  {label} {t['sentiment_label'].upper():8s}  score:{t['sentiment_score']:5.1f}  {bar}")

    ts = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")
    print(f"\n  research1 complete @ {ts}")
    print(f"  Outputs: {OUT_DIR}/tweets.csv, {OUT_DIR}/timeline.csv")

    return all_tweets, timeline

if __name__ == "__main__":
    tweets, timeline = main()
